### DISCUSSION TITLE
"Zigzagging" side sets in gmsh 2.5D meshes

### DISCUSSION THREAD

**Discussion by ln53** [idaholab/moose#24630] [Q&A Meshing] [answered]
I've been generating disjoint 2D meshes in 3D space using Gmsh. After the mesh has been processed by MOOSE (and opened in ParaView), the side sets (which should be straight along the edges) are in the pictured "zigzag" or "stitched" pattern.

- This happens using Gmsh physical groups to specify sidesets, as well as ParsedGenerateSideset, SideSetsFromNormalsGenerator, and various other side set generators.
- It only seems to happen when there are multiple disjoint meshes in 3D  - individual 2D meshes or disjoint & coplanar 2D meshes both behave as expected.
- MWE image and code under dropdowns below.

<img width="646" alt="image" src="https://github.com/idaholab/moose/assets/84389483/3aee6071-144c-451d-b1a8-f859a2c5f648">

<details>
<summary> Mesh generation with Gmsh Python API </summary>

```
import gmsh
occ = gmsh.model.occ
gmsh.initialize()

occ.addRectangle(0, 0, 0, 1, 1)
occ.addRectangle(0, 0, 0.5, 1, 1)
occ.synchronize()

gmsh.model.addPhysicalGroup(1, [1], name="MWE_Sideset")
gmsh.option.setNumber("Mesh.MeshSizeMax", 0.1)
gmsh.option.setNumber("Mesh.SaveAll", 1)
gmsh.model.mesh.generate()
gmsh.write("MWE.msh")
```

</details>

<details>
<summary> MOOSE input file mesh block </summary>

```
[Mesh]
  [mesh]
    type = FileMeshGenerator
    file = 'MWE.msh'
  []
  #   [sideset2]
  #     type = SideSetsFromNormalsGenerator
  #     input = mesh
  #     normals = '1 0 0'
  #     new_boundary = 'MWE_Sideset2'
  #   []
  #   [sideset3]
  #     type = ParsedGenerateSideset
  #     input = mesh
  #     combinatorial_geometry = 'x = 0 & z = 0'
  #     new_sideset_name = 'MWE_Sideset3'
  #   []
[]
```

</details>

**Comment by lindsayad** [comment]
This looks like fundamentally an issue in our gmsh mesh reader. @roystgnr or I will have to investigate to see what's going on, but I can't promise a fix soon

**Reply by lindsayad** [reply]
I've created an issue from this: https://github.com/idaholab/moose/issues/24631

**Reply by ln53** [reply]
Okay, thanks for the update.

**Reply by ln53** [reply]
An additional case if it helps:
- FileMeshGenerator importing a rectangular "plate" generated in Gmsh: sidesets display fine.
- Adding a GeneratedMeshGenerator plate (and combining with Gmsh plane, and translating in-plane): sidesets still display fine.
- Translating the new plate out-of-plane: sidesets display issue.
- Adding a GeneratedMeshGenerator 3D box instead of plate: sidesets display issue.

So it seems the problem is with 2D gmsh meshes and any additions to the 3rd dimension.

**Reply by roystgnr** [reply]
Check out #24631; I think the problem's with paraview.  I don't need a mesh generator to trigger it; I can do so with **vim**.

**Reply by ln53** [reply]
@roystgnr Thanks for looking into it. In that case I imagine my MOOSE models should run fine despite the odd looking sidesets in ParaView? I will raise this on the ParaView forum instead!

**Reply by roystgnr** [reply]
> In that case I imagine my MOOSE models should run fine despite the odd looking sidesets in ParaView?

Yes, but remember that "convergence studies" and "manufactured solutions" are standard subcategories of "verification", whereas "imagination" is not.  ;-)  If you're writing new models you should generally be double-checking your convergence rate against a benchmark problem in any case; if you're just assembling existing MOOSE kernels then normally I'd say you can trust us but this might be a case where trust-but-verify is a better philosophy.

**Reply by ln53** [reply]
Indeed! We're using existing shell/solid mechanics kernels and doing some validation against NASTRAN - I hadn't even bothered running the models when I saw this issue, but now I know it's only a ParaView artefact I'll press on. Thanks again.

**Reply by roystgnr** [reply]
There appears to be a little model of Tinsley Oden living in my head, who insists on pointing out that code-to-code comparisons are still just a "verification" ("solving the equations right") subcategory; it's not "validation" ("solving the right equations") until you're comparing to experimental data for your problem too.

**Reply by ln53** [reply]
@roystgnr update on this, it seems Paraview is behaving as expected (albeit a little confusingly) - here's the relevant excerpt from the [ParaView thread](https://discourse.paraview.org/t/zigzagging-jagged-sidesets-with-2d-out-of-plane-triangular-meshes/12320/10?u=ln53):

> From Greg Sjaardema, the designer and developer of the Exodus format:
> 
> OK, took me too long to figure out what is happening. For a 4-noded element, we have a quad in 2D and a shell in 3D. However, when we started using 3-noded elements more, we decided for some reason to name both the 3D shell version and the 2D planar version “triangle” or “tri”.
> 
> We then determine whether it is a normal 3-sided two-dimensional triangular element, or a “5-sided” triangular shell element based on the dimensionality of the database.
> 
> In this case, we have a three dimensional database, so the “tri” elements in the element block are “triangular shell” elements.
> 
> Given a triangular shell with connectivity <1 2 3>, we have:
> 
> Face 1 is triangle front face with connectivity 1 2 3
> Face 2 is triangle back face with reverse connectivity 3 2 1
> Face 3 is first edge with connectivity 1 2. [ This is face 1 in a 2d triangle element]
> Face 4 is second edge – 2 3
> Face 5 is third edge – 3 1
> The mesh you have is assuming that the elements are 3-sided elements and the “sides” are all edges, so the side numbering is off by 2. — They should have all sides as ‘5’ instead of ‘3’ to get the sideset definition that they think they have.
> 
> Hope that makes sense.
> 
> …Greg

This is outlined under the "Triangular" section of the Exodus II element types [documentation](https://sandialabs.github.io/seacas-docs/html/element_types.html), suggesting MOOSE is off by 2 when numbering the edges. I removed Gmsh from the equation just to be sure, and managed to produce a similar issue in `MWE_2` below. This should add the triangle edges along the mesh boundary to the sideset, but selects some of the triangle faces and at least one "jaggy" edge.

<img width="772" alt="image" src="https://github.com/idaholab/moose/assets/84389483/4ae28bf3-b189-4bd6-b08a-ab718673e17a">


<details>
<summary>MWE_2.i</summary>

```
[Mesh]
  [poly]
    type = PolyLineMeshGenerator
    points = '0 0 0
              0 1 0
              1 1 0
              1 0 0'
    loop = true
  []
  [triangulate]
    type = XYDelaunayGenerator
    boundary = 'poly'
    add_nodes_per_boundary_segment = 9
    refine_boundary = false
    desired_area = 0.007
  []
  [duplicate]
    type = CombinerGenerator
    inputs = 'triangulate'
    positions = '0 0 0 0 0 0.5'
  []
  [sideset]
    type = ParsedGenerateSideset
    input = 'duplicate'
    combinatorial_geometry = 'z = 0 & x = 0'
    new_sideset_name = 'mwe_sideset'
  []
[]
```

</details>

**Reply by roystgnr** [reply]
Well, that's horrifying.  I'll see if I can fix it in our ExodusII_IO.

**Reply by ln53** [reply]
Excellent, thank you!

**Reply by roystgnr** [reply, marked as answer]
The fix is finally in MOOSE.  Sorry that took so long!

**Reply by ln53** [reply]
No problem, thanks - appreciate all the work that went into it!
